### PR TITLE
Adding TFv2 support for Faster-RCNN

### PR DESCRIPTION
**Pull request by kieranfraser**
# Description

This PR introduces TensorFlow v2 support for the [Faster-RCNN object detection model](https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/tf2_detection_zoo.md). 

Includes:
- TensorFlowV2FasterRCNN estimator
- Tests for estimator in test_tensorflow_v2_faster_rcnn.py
- Example notebook "estimator_tf2_faster_rcnn.ipynb" detailing installation and usage

Fixes # (issue)
#1616

## Type of change

Please check all relevant options.

- [x] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [ ] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [x] Test object detection
- [x] Test compute gradients and losses
- [x] Test errors thrown on invalid parameters (defences/clipping)

**Test Configuration**:
- OS
- Python version
- ART version or commit number
- TensorFlow / Keras / PyTorch / MXNet version

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/estimators/object_detection/tensorflow_v2_faster_rcnn.py [hunk 1]
```diff
@@ -0,0 +1,490 @@
...
+        of (predictions, losses, detections).
+
+        :param filename: Name of the file.
+        :param url: Download URL.
+        :param obj_detection_model: The model
+        :param is_training: A boolean indicating whether the training version of the computation graph should be
+                            constructed.
+        :return: the object detection model restored from checkpoint
+        """
+        import tensorflow as tf  # lgtm [py/repeated-import]
```
Comment (beat-buesser) [inline art/estimators/object_detection/tensorflow_v2_faster_rcnn.py:181]:
```suggestion
        import tensorflow as tf
```

File: art/estimators/object_detection/tensorflow_v2_faster_rcnn.py [hunk 2]
```diff
@@ -0,0 +1,490 @@
...
+                 - labels [N]: the labels for each image in TensorFlow (`standardise_output=False`) or PyTorch
+                               (`standardise_output=True`) format
+
+        :param standardise_output: True if `y` is provided in standardised PyTorch format. Box coordinates will be
+                                   scaled back to [0, 1], label index will be decreased by 1 and the boxes will be
+                                   changed from [x1, y1, x2, y2] to [y1, x1, y2, x2] format, with
+                                   0 <= x1 < x2 <= W and 0 <= y1 < y2 <= H.
+        :return: Loss gradients of the same shape as `x`.
+        """
+        import tensorflow as tf  # lgtm [py/repeated-import]
```
Comment (beat-buesser) [inline art/estimators/object_detection/tensorflow_v2_faster_rcnn.py:239]:
```suggestion
        import tensorflow as tf
```

File: art/estimators/object_detection/tensorflow_v2_faster_rcnn.py [hunk 3]
```diff
@@ -0,0 +1,490 @@
...
+                 fields of the Dict are as follows:
+
+                 - boxes [N, 4]: the boxes in [y1, x1, y2, x2] format, with 0 <= x1 < x2 <= W and 0 <= y1 < y2 <= H.
+                                 Can be changed to PyTorch format with `standardise_output=True`.
+                 - labels [N]: the labels for each image in TensorFlow format. Can be changed to PyTorch format with
+                               `standardise_output=True`.
+                 - scores [N]: the scores or each prediction.
+        """
+
+        import tensorflow as tf  # lgtm [py/repeated-import]
```
Comment (beat-buesser) [inline art/estimators/object_detection/tensorflow_v2_faster_rcnn.py:317]:
```suggestion
        import tensorflow as tf
```

File: art/estimators/object_detection/tensorflow_v2_faster_rcnn.py [hunk 4]
```diff
@@ -0,0 +1,490 @@
...
+        :param x: Sample input with shape as expected by the model.
+        :param y: Targets of format `List[Dict[str, np.ndarray]]`, one for each input image. The fields of the Dict are
+                  as follows:
+                    - boxes [N, 4]: the boxes in [y1, x1, y2, x2] format, with 0 <= x1 < x2 <= W and 0 <= y1 < y2 <= H.
+                    - labels [N]: the labels for each image in TensorFlow format.
+                    - scores [N]: the scores or each prediction.
+        :return: np.float32 representing total loss.
+        """
+
+        import tensorflow as tf  # lgtm [py/repeated-import]
```
Comment (beat-buesser) [inline art/estimators/object_detection/tensorflow_v2_faster_rcnn.py:411]:
```suggestion
        import tensorflow as tf
```

File: tests/estimators/object_detection/test_tensorflow_v2_faster_rcnn.py [hunk 5]
```diff
@@ -0,0 +1,336 @@
+# MIT License
+#
+# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2020
```
Comment (beat-buesser) [inline tests/estimators/object_detection/test_tensorflow_v2_faster_rcnn.py:3]:
```suggestion
# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2022
```

File: tests/estimators/object_detection/test_tensorflow_v2_faster_rcnn.py [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import logging
+
+import numpy as np
+import pytest
+from art.utils import load_dataset
```
Comment (beat-buesser) [inline tests/estimators/object_detection/test_tensorflow_v2_faster_rcnn.py:24]:
```suggestion
import pytest

from art.utils import load_dataset
```

File: art/estimators/object_detection/tensorflow_v2_faster_rcnn.py [hunk 7]
```diff
@@ -0,0 +1,490 @@
+# MIT License
```
Comment (beat-buesser) [inline art/estimators/object_detection/tensorflow_v2_faster_rcnn.py:1]: Please rename this file to `tensorflow_faster_rcnn.py`

File: art/estimators/object_detection/tensorflow_v2_faster_rcnn.py [hunk 8]
```diff
@@ -0,0 +1,490 @@
...
+            # If obj_detection_model is None, then we need to have parameters filename and url to download, extract
+            # and load the object detection model
+            if filename is None or url is None:  # pragma: no cover
+                raise ValueError(
+                    "Need input parameters `filename` and `url` to download, "
+                    "extract and load the object detection model."
+                )
+
+            # Download and extract
+            print("downloading file")
```
Comment (beat-buesser) [inline art/estimators/object_detection/tensorflow_v2_faster_rcnn.py:196]: Please use logging instead of print.

File: art/estimators/object_detection/tensorflow_v2_faster_rcnn.py [hunk 9]
```diff
@@ -0,0 +1,490 @@
...
+            path = get_file(filename=filename, path=config.ART_DATA_PATH, url=url, extract=True)
+            print("loading model config")
```
Comment (beat-buesser) [inline art/estimators/object_detection/tensorflow_v2_faster_rcnn.py:198]: Please use logging instead of print.

File: art/estimators/object_detection/tensorflow_v2_faster_rcnn.py [hunk 10]
```diff
@@ -0,0 +1,490 @@
...
+            ].faster_rcnn.first_stage_max_proposals
+
+            # Load model
+            obj_detection_model = model_builder.build(
+                model_config=configs["model"], is_training=is_training, add_summaries=False
+            )
+
+        # Restore checkpoint
+        ckpt = tf.compat.v2.train.Checkpoint(model=obj_detection_model)
+        ckpt.restore(path + "/checkpoint/ckpt-0").expect_partial()
```
Comment (beat-buesser) [inline art/estimators/object_detection/tensorflow_v2_faster_rcnn.py:213]: How is `path` defined if obj_detection_model is not None?

File: art/estimators/object_detection/tensorflow_v2_faster_rcnn.py [hunk 11]
```diff
@@ -0,0 +1,490 @@
...
+                    from art.estimators.object_detection.utils import convert_tf_to_pt
+
+                    d_sample = convert_tf_to_pt(y=[d_sample], height=x.shape[1], width=x.shape[2])[0]
+
+                d_sample["scores"] = batch_results["detection_scores"][i].numpy()
+
+                results.append(d_sample)
+
+        self._detections = results
+        self._predictions = [i["scores"] for i in results]
```
Comment (beat-buesser) [inline art/estimators/object_detection/tensorflow_v2_faster_rcnn.py:357]: Please add definitions/initialisations for the attributes in __init__.

File: art/estimators/object_detection/tensorflow_v2_faster_rcnn.py [hunk 12]
```diff
@@ -0,0 +1,490 @@
...
+            groundtruth_boxes_list=groundtruth_boxes_list,
+            groundtruth_classes_list=groundtruth_classes_list,
+            groundtruth_weights_list=groundtruth_weights_list,
+        )
+
+        preprocessed_images, true_image_shapes = self._model.preprocess(x_preprocessed)
+        predictions = self._model.predict(preprocessed_images, true_image_shapes)
+        losses = self._model.loss(predictions, true_image_shapes)
+
+        self._losses: Dict[str, np.ndarray] = {}
```
Comment (beat-buesser) [inline art/estimators/object_detection/tensorflow_v2_faster_rcnn.py:486]: Please add definitions/initialisation for the attributes in \_\_init__.

**Comment by beat-buesser** [comment]
- Please create a similar file as https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/main/.github/workflows/ci-pytorch-object-detectors.yml to run the new tests.
- Please update `notebooks/README.md` with the new notebook

**Review by beat-buesser — CHANGES_REQUESTED** [review]
Hi @kieranfraser Thank you for your pull request! I have added my review, what do you think?

File: art/estimators/object_detection/tensorflow_v2_faster_rcnn.py (hunk 9 above)
Comment (kieranfraser) [reply art/estimators/object_detection/tensorflow_v2_faster_rcnn.py]: I removed this as I added for debugging purposes.

File: art/estimators/object_detection/tensorflow_v2_faster_rcnn.py (hunk 10 above)
Comment (kieranfraser) [reply art/estimators/object_detection/tensorflow_v2_faster_rcnn.py:206]: Good catch - this should only occur when model is None! If the user passes an existing model, they should already have restored from checkpoint.

**Review by beat-buesser — APPROVED** [review]
Hi @kieranfraser Thank you very much for contributing support for object detection models in TensorFlow v2 to ART!
